docs: add Kafka ADRs (#11)
edx/edx-arch-experiments: edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst

=====================
Kafka-Based Event Bus
=====================

Status
------

Provisional

Context
-------

The draft `OEP-52: Event Bus Architecture`_ explains how the Open edX platform would benefit from an event bus, as well as providing some additional decisions around the event bus. One decision is to enable the event bus technology to be pluggable through some abstraction layer.

> However, an initial technology, even if multiple could eventually be selected, is still required.

Reviewer: The construction of this sentence is a little hard to follow

```suggestion
This still requires selecting specific technology for the first implementation.
```

Author: Thanks. I'll use this, but add `a` before `specific technology`.

.. _`OEP-52: Event Bus Architecture`: https://github.com/openedx/open-edx-proposals/pull/233

Decision
--------

An initial implementation of an event bus for the Open edX platform will be implemented using `Kafka`_. This implementation will be used by edx.org, and available to the Open edX community.

> However, given the intention of creating an abstraction layer around the event bus technology, this decision does not preclude the introduction of alternative event bus implementations based on other technologies in the future.

Reviewer: I think the first clause unnecessarily complicates the sentence since it's not really a contradiction.

```suggestion
This decision does not preclude the introduction of alternative event bus implementations based on other technologies in the future.
```

Author: 👍

.. _Kafka: https://kafka.apache.org/

Why Kafka?
~~~~~~~~~~

Kafka is a distributed streaming platform. Kafka's implementation maps nicely to the pub/sub pattern. However, some native features of a message broker are not built-in.

Kafka has been around for a long time. See `Thoughtworks's technology radar introduced Kafka`_ as "Assess" in 2015, and "Trial" in 2016. It never moved up to "Adopt", and also never moved down to "Hold". Read `Thoughtwork's Kafka decoder page`_ to learn more about its benefits and trade-offs, and how it is used.

> More recently, the `Thoughtworks's technology radar introduced Apache Pulsar`_ as "assess" in 2020, and the `technology radar introduced Kafka API without Kafka`_ in 2021. This both demonstrates the de facto standard of the Kafka API, but also Thoughtwork's hope to find a less complex alternative.

Reviewer: First sentence is missing a verb.

Author: Do you mean the sentence beginning with "More recently"? The verb is "introduced", and appears twice, although both times in the midst of a link, which may be why it was missed?

We believe Apache Kafka is still the right option due to its maturity, documentation, support and community.

.. _Thoughtworks's technology radar introduced Kafka: https://www.thoughtworks.com/radar/tools/apache-kafka
.. _Thoughtwork's Kafka decoder page: https://www.thoughtworks.com/decoder/kafka

.. _Thoughtworks's technology radar introduced Apache Pulsar: https://www.thoughtworks.com/radar/platforms/apache-pulsar
.. _technology radar introduced Kafka API without Kafka: https://www.thoughtworks.com/radar/platforms/kafka-api-without-kafka

Kafka Highlights
~~~~~~~~~~~~~~~~

Pros
^^^^

* Battle-tested, widely adopted, big community, lots of documentation and answers.
* Enables event replay-ability.

Cons
^^^^

* Complex to manage, including likely manual scaling.

> * Simple consumers requires additional code for some messaging features.

Reviewer: s/requires/require/

Author: 👍

Consequences
------------

> * Operators will need to deploy and manage the selected infrastructure, which is likely to be complex. If Apache Kafka is selected, there are likely to be a set of auxiliary parts to provide all required functionality for our message bus. However, third-party hosting is also available (see separate decision).

Reviewer: Nit: 'which will likely be' is a little more compact a construction than 'which is likely to be'

> * Most of the consequences of an event bus should related to `OEP-52: Event Bus Architecture`_ more generally, and hopefully will not be Kafka specific.

Reviewer: s/related/relate/

Rejected Alternatives
---------------------

Apache Pulsar
~~~~~~~~~~~~~

> Although rejected to start, `Apache Pulsar`_ remains an option for those looking for an alternative to Kafka.

Reviewer: ```suggestion
Although rejected for initial edx.org implementation, `Apache Pulsar`_ remains an option for those looking for an alternative to Kafka.
```

Pros
^^^^

* Ease of scalability (built-in, according to docs).

> * Ease of data retention capabilities.

Reviewer: 'ease of capabilities' is a strange construction. Maybe  `good data retention capabilities` or `easy data retention`

* Additional built-in pub/sub features (built-in, according to docs).

Cons
^^^^

* Requires 3rd party hosting or larger upfront investment if self-hosted (kubernetes).
* Less mature (but growing) community, little documentation, and few answers.
* Python built-in schema management is buggy and hard to work with for complex use cases.

Note: Read an interesting (Kafka/Confluent) biased article exploring `comparisons and myths of Kafka vs Pulsar`_.

.. _Apache Pulsar: https://pulsar.apache.org/
.. _comparisons and myths of Kafka vs Pulsar: https://dzone.com/articles/pulsar-vs-kafka-comparison-and-myths-explored

Redis
~~~~~

Pros
^^^^

> * Already part of Open edX platform

Reviewer: Nit: I don't like using periods for sentence fragments but if we use them everywhere else we might as well be consistent : P
```suggestion
* Already part of Open edX platform.
```

Cons
^^^^

* Can lose acked data, even if RAM backed up with an append-only file (AOF).
* Requires homegrown schema management.

RabbitMQ
~~~~~~~~

Pros
^^^^

* Built-in message broker capabilities like routing, filtering, and fault handling.

Cons
^^^^

* Not built for message retention or message ordering.

AWS SNS/SQS
~~~~~~~~~~~

Pros
^^^^

* Simpler hosting for those self-hosting in AWS.

Cons
^^^^

* Cannot be shared as an open source solution.
* Events are not replayable.

Additional References
---------------------

> * Technology comparisons performed by edX.org

Reviewer: Nit
```suggestion
* Technology comparisons performed by edX.org:
```
